support-frontend/app/actions/AsyncAuthenticatedBuilder.scala

```diff
@@ -0,0 +1,41 @@
+package actions
+
+import com.typesafe.scalalogging.LazyLogging
+import play.api.mvc.Security.AuthenticatedRequest
+import play.api.mvc._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+// scalastyle:off
+// This is based on AuthenticationBuilder in https://github.com/playframework/playframework/blob/1ca9d0af237ac45a011671b2e036d726cd05e4e7/core/play/src/main/scala/play/api/mvc/Security.scala
+// The main difference is that this action builder enables authenticating a user asynchronously.
+// i.e. RequestHeader => Future[Option[U]] instead of RequestHeader => Option[U]
+// In the context of Guardian applications, asynchronous authentication equals making a call to identity API,
+// to authenticate the user data.
+// TODO: considering porting this to identity-play-auth.
+// scalastyle:on
+class AsyncAuthenticatedBuilder[U](
+    userinfo: RequestHeader => Future[Option[U]],
+    defaultParser: BodyParser[AnyContent],
+    onUnauthorized: RequestHeader => Result
+)(implicit val executionContext: ExecutionContext)
+  extends ActionBuilder[({ type R[A] = AuthenticatedRequest[A, U] })#R, AnyContent] with LazyLogging { // scalastyle:ignore
```

Comment:
good old type lambdas! - consider pull this out as a named type to make it less WTF (I have experience of this situation!)

Reply:
I can't remember exactly and I'm now wondering if it's possbile in all cases but something like 
```scala
class A[U] {
  type R[A] = AuthenticatedRequest[A, U]
}
```
then
```scala
... ActionBuilder[A[U]#R, AnyContent] ...
````

Reply:
I'm a but confused on how you'd pull it out since it's reliant on `U` which is a type parameter for the class. I've used [typeleve/kind-projector](https://github.com/typelevel/kind-projector) in other applications, which I'd be happy to use here to simplify the syntax (?)

Reply:
maybe it's not possible, as I was typing it all out I was starting to have doubts about scoping! sorry for the red herring 🐠